cli/src/cmd/forge/script/receipts.rs

```diff
@@ -1,90 +1,78 @@
-use crate::{cmd::ScriptSequence, utils::print_receipt};
-use ethers::prelude::{Http, Middleware, Provider, TransactionReceipt, H256, U256};
-use futures::future::join_all;
+use std::sync::Arc;
 
-use super::broadcast::BroadcastError;
+use crate::{cmd::ScriptSequence, init_progress, update_progress, utils::print_receipt};
+use ethers::prelude::{Http, PendingTransaction, Provider, RetryClient, TxHash};
+use futures::StreamExt;
+use indicatif::{ProgressBar, ProgressStyle};
 
 /// Gets the receipts of previously pending transactions.
 pub async fn wait_for_pending(
-    provider: &Provider<Http>,
+    provider: Arc<Provider<RetryClient<Http>>>,
     deployment_sequence: &mut ScriptSequence,
 ) -> eyre::Result<()> {
     if !deployment_sequence.pending.is_empty() {
         println!("##\nChecking previously pending transactions.");
-        let future_receipts = deployment_sequence
-            .pending
-            .iter()
-            .map(|tx_hash| pending_receipt(provider, *tx_hash))
-            .collect();
-        wait_for_receipts(future_receipts, deployment_sequence).await?;
+        wait_for_receipts(deployment_sequence.pending.clone(), deployment_sequence, provider)
+            .await?;
     }
     Ok(())
 }
 
-/// Waits for a pending receipt, and gets its nonce to return (receipt, nonce).
-async fn pending_receipt(
-    provider: &Provider<Http>,
-    tx_hash: H256,
-) -> Result<(TransactionReceipt, U256), BroadcastError> {
-    let pending_err =
-        || BroadcastError::Simple(format!("Failed to get pending transaction {tx_hash:?}."));
-
-    let receipt = provider
-        .get_transaction_receipt(tx_hash)
-        .await
-        .map_err(|_| pending_err())?
-        .ok_or_else(pending_err)?;
-
-    let tx = provider
-        .get_transaction(tx_hash)
-        .await
-        .map_err(|_| pending_err())?
-        .ok_or_else(pending_err)?;
-
-    Ok((receipt, tx.nonce))
-}
-
 /// Waits for a list of receipts. If it fails, it tries to retrieve the transaction hash that can be
 /// used on a later run with `--resume`.
 pub async fn wait_for_receipts(
-    tasks: Vec<impl futures::Future<Output = Result<(TransactionReceipt, U256), BroadcastError>>>,
+    tx_hashes: Vec<TxHash>,
     deployment_sequence: &mut ScriptSequence,
+    provider: Arc<Provider<RetryClient<Http>>>,
 ) -> eyre::Result<()> {
-    let res = join_all(tasks).await;
+    let mut tasks = futures::stream::iter(
+        tx_hashes.iter().map(|tx| PendingTransaction::new(*tx, &provider)).collect::<Vec<_>>(),
+    )
+    .buffered(2);
```

Comment:
Why 2? Should we turn that up a bit? Maybe `tx_hashes.len() / [some constant]` or something?

Reply:
we can definitely go a bid wilder here, since these are just requested to retrieve receipts. so perhaps we can do something like `tx_hashes.len().max(xx)` or something.

Reply:
Put it at 10 and it does feel fast (if they're mined obviously). More than that I was hitting some hangs, probably coming from the exponential back-off.